Add TLS certs support for mutual authentication

zhiburt opened the pull request:
Hi there

Thanks for the `good first issue` label

Let me know if I did something wrong (which I could :disappointed:).

closes #82

kursataktas commented on core/types/scenario.go:
```diff
@@ -102,6 +105,12 @@ type ScenarioItem struct {
 	// Authentication
 	Auth Auth
 
+	// A TLC cert
```
TLC -> TLS

kursataktas commented on core/types/scenario.go:
```diff
@@ -102,6 +105,12 @@ type ScenarioItem struct {
 	Auth Auth
 
+	// A TLC cert
+	Cert tls.Certificate
+
+	// A TLC cert pool
```
same ⬆️

kursataktas commented on core/types/scenario.go:
```diff
@@ -167,6 +176,27 @@ func (si *ScenarioItem) validate() error {
 	return nil
 }
 
+func (si *ScenarioItem) ParseTLC(certFile, keyFile string) error {
```
**_Nice to have;_**

What about changing the contract to below;
`func ParseTLC(certFile, keyFile string) (string, string, error)`

We can return `(cert, certpool, err)` if suitable, `(nil, nil, err)` if not suitable. Then we don't need the `if` statements both on `main.go` and` json.go`:
https://github.com/zhiburt/ddosify/blob/patch-tls-certs/main.go#L252
https://github.com/zhiburt/ddosify/blob/patch-tls-certs/config/json.go#L151

 WDYT?

zhiburt replied:
Well... you will then ignore error?
Or you will have to introduce a custom error type to recognize if it failed because the path was empty or there was some issue.

I agree that the interface of the function maybe not the best (see how it used it tests :disappointed_relieved:).
But what I am saying the check will not go away right (because of what I've pointed out above)?

kursataktas replied:
No need to discard error logic. Please take a look `AdjustUrlProtocol`. We can apply the same strategy. As a said it is nice to have for now.

zhiburt replied:
https://github.com/ddosify/ddosify/blob/3db5d92513e426998badb619ecc783135d84584e/core/types/scenario.go#L179

kursataktas commented on core/scenario/requester/http.go:
```diff
@@ -317,6 +317,12 @@ func (h *HttpRequester) initTLSConfig() *tls.Config {
 		InsecureSkipVerify: true,
 	}
+
+	if h.packet.CertPool != nil {
```
Let's check `h.packet.Cert` too. Just in case.

kursataktas commented on core/engine_test.go:
```diff
@@ -624,3 +627,178 @@ func TestDynamicData(t *testing.T) {
+	e.Start()
+
+	if handlerCalls != 0 {
+		t.Errorf("handler was called uncexpectedly: %#v", handlerCalls)
```
typo --> `uncexpectedly`

kursataktas reviewed (commented):
Thank you for this fantastic work 🎊 🚀 

I left a few comments. Please take a look at them.

Also, you may consider adding some test cases on `json_test.go` and `main_test.go`

zhiburt commented:
> Also, you may consider adding some test cases on json_test.go and main_test.go

done

kursataktas commented on main_test.go:
```diff
@@ -239,6 +250,89 @@ func TestCreateScenario(t *testing.T) {
+				if err != nil {
+					t.Errorf("Errored: %v", err)
+				}
+				if reflect.DeepEqual(test.expected, s) {
```
this should be `!reflect.DeepEqual(test.expected, s)`.
Looks like same bug happened on  `TestCreateScenario`. I'll send a fix for it on another PR.

kursataktas replied:
Here is the fixed version of `TestCreateScenario`.  If you would like, you can send this fix along with this PR.

```
func TestCreateScenario(t *testing.T) {
	url := "https://test.com"
	valid := types.Scenario{
		Scenario: []types.ScenarioItem{
			{
				ID:       1,
				Protocol: types.DefaultProtocol,
				Method:   types.DefaultMethod,
				URL:      url,
				Timeout:  types.DefaultTimeout,
				Headers:  map[string]string{},
			},
		},
	}
	validWithAuth := types.Scenario{
		Scenario: []types.ScenarioItem{
			{
				ID:       1,
				Protocol: types.DefaultProtocol,
				Method:   types.DefaultMethod,
				URL:      url,
				Timeout:  types.DefaultTimeout,
				Headers:  map[string]string{},
				Auth: types.Auth{
					Type:     types.AuthHttpBasic,
					Username: "testuser",
					Password: "pass",
				},
			},
		},
	}

	tests := []struct {
		name      string
		args      []string
		shouldErr bool
		expected  types.Scenario
	}{
		{"InvalidAuth", []string{"-t=https://test.com", "-a=no_pass_included"}, true, types.Scenario{}},
		{"InvalidTarget", []string{"-t=asds.x.x.x"}, true, types.Scenario{}},
		{"Valid", []string{"-t=https://test.com"}, false, valid},
		{"ValidWithAuth", []string{"-t=https://test.com", "-a=testuser:pass"}, false, validWithAuth},
	}

	for _, test := range tests {
		tf := func(t *testing.T) {
			// Arrange
			resetFlags()
			oldArgs := os.Args
			defer func() {
				os.Args = oldArgs
			}()

			os.Args = []string{"cmd"}
			for _, a := range test.args {
				os.Args = append(os.Args, a)
			}

			// Act
			flag.Parse()
			s, err := createScenario()

			// Assert
			if test.shouldErr {
				if err == nil {
					t.Errorf("Should be errored")
				}
			} else {
				if err != nil {
					t.Errorf("Errored: %v", err)
				}
				if !reflect.DeepEqual(test.expected, s) {
					t.Errorf("Expected %#v, Found %#v", test.expected, s)
				}
			}

		}

		t.Run(test.name, tf)
	}
}
```

kursataktas replied:
I sent the fix

kursataktas replied:
@zhiburt I think this issue still exists. I fixed the `TestCreateScenario`. `TestCreateScenarioTLS` still has incorrect comparison.

zhiburt replied:
fixed

kursataktas commented:
TODO: Readme update

zhiburt commented on main.go:
```diff
@@ -235,22 +238,31 @@ func createScenario() (s types.Scenario, err error) {
+		scenario.CertPool = pool
+	}
+
+	*protocol = strings.ToUpper(*protocol)
```
Good catch. Sorry...

kursataktas commented:
Everything looks clear @zhiburt 🚀 

There are 2 things left before merging this branch:

1. Could you please update the [related section of the Readme](https://github.com/ddosify/ddosify#cli-flags)
2. Could you please update the description of this PR with the explanation of your manual test steps? It would be useful for future bug investigations.

zhiburt commented:
> Could you please update the [related section of the Readme](https://github.com/ddosify/ddosify#cli-flags)

Added

> Could you please update the description of this PR with the explanation of your manual test steps? It would be useful for future bug investigations.

I don't have any manual steps :)
I've added an descriptions for the tests in `engine_test.go`.

To test it manually you would need to spawn a webserver with chosen certs and direct ddosify to the webserver.

zhiburt commented:
I've played a little bit and it works.
Based on https://medium.com/@sevcsik/authentication-using-https-client-certificates-3c9d270e8326

1. Generate certificates following the guide.
2. Run this server `node your_file.js`

```js
const https = require('https')
const fs = require('fs');

let server = https.createServer({
    requestCert: true,
    rejectUnauthorized: false,
    key: fs.readFileSync('server_key.pem'),
    cert: fs.readFileSync('server_cert.pem'),
    ca: [ fs.readFileSync('server_cert.pem') ]
},
(req, res)  => {
    const cert = req.connection.getPeerCertificate()
    if (cert != null && cert.subject != null) {
        console.info(`Hello ${cert.subject.CN}, your certificate was issued by ${cert.issuer.CN}!`)
    }

    if (!req.client.authorized) {
        res.writeHead(401);
        return res.end('Invalid client certificate authentication.');
      }

      console.info("123213321");

      res.writeHead(200);
      res.end('Hello, world!');
});

server.listen(8098)
```

3. Run ddosify without certs


```bash
go run main.go -t https://127.0.0.1:8098
```

![image](https://user-images.githubusercontent.com/20165848/194963615-29cbcbbb-395b-4b7c-a47b-c6c9d46a7f1a.png)

4. Run ddosify with alice certs

```bash
go run main.go -t https://127.0.0.1:8098 --cert_path=/alice_cert.pem --cert_key_path=/alice_key.pem
```

![image](https://user-images.githubusercontent.com/20165848/194963643-b3fb338b-f24b-4b91-b4b7-8da30b0dd51c.png)

kursataktas commented:
Fascinating work @zhiburt 🎊  Thank you very much for your effort. I'll merge this PR this week.
